### PR TITLE
[ICONS] Refactor to not auto-include the whole set when importing one icon

### PR DESCRIPTION
**Pull request by jeeyun**
With this refactor, importing one icon will not result in adding all shapes from the set as per #2599 .

Now we have `-umd.js` and `-umd.min.js` files for each shape, that are meant to be included in script tags in static sites. Including those scripts will result in adding all shapes from that file.

To summarize usage:
## Angular clients using Typescript

```
import { ClarityIcons } from '@clr/icons';
import { EssentialShapes } from '@clr/icons/shapes/essential';
import { ClrShapeStar } from '@clr/icons/shapes/social';

ClarityIcons.init(); // this should be done once at some root level of the application
ClarityIcons.add(EssentialShapes); // adds the whole set
ClarityIcons.add(ClrShapeStar); // adds a single icon from this set
```

## Static sites loading javascript files via script tag
To include all icons:
```
<script src="clr-icons/clr-icons-all-shapes.min.js"></script>
```
Or to include specific icon sets:
```
<script src="clr-icons/clr-icons-base.min.js"></script>
<script src="clr-icons/shapes/core-umd.min.js"></script>
<script src="clr-icons/shapes/trave-umd.min.js"></script>
```

Signed-off-by: Jeeyun Lim <jeeyun.lim@gmail.com>

### CODE REVIEW

File: src/clr-icons/shapes/chart-shapes-umd.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { ChartShapes, ClrShapeLineChartSVG } from './chart-shapes';
+import { descriptorConfig } from '../utils/descriptor-config';
+
+/* tslint:disable:variable-name */
```
[the remark this answers is not in the document]
Comment (jeeyun) [reply src/clr-icons/shapes/chart-shapes-umd.ts]: The file name changes in #2 make sense to me. For #1, are you suggesting that we change the name in both ts and js versions? So our imports will be: 

```import { EssentialShapes } from '@clr/icons/shapes/essential';```

And our script tags:

```<script src="clr-icons/shapes/essential-umd.min.js"></script>```

[the remark this answers is not in the document]
Comment (gnomeontherun) [reply src/clr-icons/shapes/chart-shapes-umd.ts]: Correct, seems simpler to me.

File: scripts/clr-icons-svg.js [hunk 2]
```diff
@@ -7,19 +7,10 @@
...
-  'core-shapes',
-  'commerce-shapes',
-  'essential-shapes',
-  'media-shapes',
-  'social-shapes',
-  'technology-shapes',
-  'travel-shapes',
-  'chart-shapes',
-];
+const SHAPE_SETS = ['core', 'commerce', 'essential', 'media', 'social', 'technology', 'travel', 'chart'];
```
Comment (Shijir) [inline scripts/clr-icons-svg.js]: Could you add `text-edit-shapes` here?

File: src/clr-icons/clr-icons-api.ts [hunk 3]
```diff
@@ -63,6 +64,16 @@ export class ClarityIconsApi {
     }
   }
 
+  init(options: any = { includeCore: true }) {
```
Comment (Shijir) [inline src/clr-icons/clr-icons-api.ts]: I think it's better to use an interface here instead of this `any` type.

File: scripts/clr-icons-svg.js (hunk 2 above)
Comment (jeeyun) [reply scripts/clr-icons-svg.js]: Thanks for catching that. Should be fixed now.

File: src/clr-icons/clr-icons-api.ts (hunk 3 above)
Comment (jeeyun) [reply src/clr-icons/clr-icons-api.ts]: Yeah. We made a change to actually pass in any set optionally for init(). You can take a look at the latest changes. The reason for this change is.. if we explicitly have a shapeSet like the way I had it initially it doesn't tree-shake the way we want it to.

File: scripts/write-svg-icons.js [hunk 4]
```diff
@@ -177,12 +177,14 @@ let makeSVG = (shapeTitle, shapeContent) => {
...
-  // will convert core-shapes into CoreShapes
-  return kebabCase
-    .split('-')
-    .map(splitPart => splitPart.charAt(0).toUpperCase() + splitPart.slice(1))
-    .join('');
+let convertToExportedName = setName => {
+  // will convert core into CoreShapes
+  return (
+    setName
+      .split('-')
```
Comment (gnomeontherun) [inline scripts/write-svg-icons.js]: I don't think we need to split anymore, just uppercase the first letter and append `Shapes`.

File: src/clr-icons/api-autoload.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { ClarityIconsApi } from './api';
```
Comment (gnomeontherun) [inline src/clr-icons/api-autoload.ts]: Let's rename this to `api-autoload.umd.ts`for consistency with naming umd files.

File: src/clr-icons/clr-icons.spec.ts [hunk 6]
```diff
@@ -35,18 +35,23 @@ import { clrIconSVG } from './utils/svg-tag-generator';
...
+  { name: 'Commerce shapes', shapes: CommerceShapes, randomShape: ClrShapeECheck },
+  { name: 'Essential shapes', shapes: EssentialShapes, randomShape: ClrShapePencil },
+  { name: 'Social shapes', shapes: SocialShapes, randomShape: ClrShapeStar },
+  { name: 'Media shapes', shapes: MediaShapes, randomShape: ClrShapePlay },
+  { name: 'Travel shapes', shapes: TravelShapes, randomShape: ClrShapeCar },
+  { name: 'Technology shapes', shapes: TechnologyShapes, randomShape: ClrShapeCPU },
+  { name: 'Chart shapes', shapes: ChartShapes, randomShape: ClrShapeBarChart },
+  { name: 'Text Edit shapes', shapes: TextEditShapes, randomShape: ClrShapeBold },
 ];
 /* tslint:enable:no-string-literal */
```
Comment (gnomeontherun) [inline src/clr-icons/clr-icons.spec.ts]: Can we add a test case scenario where we don't run `init()` and to see that it doesn't set the global object?

File: src/clr-icons/clr-icons.spec.ts [hunk 7]
```diff
@@ -391,7 +399,7 @@ describe('ClarityIcons', () => {
 
       const clrIconUniqId = clarityIcon.clrIconUniqId;
       const testShape = giveAngleShapeTitle(clrIconUniqId, customTitle);
-
+      console.log(clarityIcon);
```
Comment (gnomeontherun) [inline src/clr-icons/clr-icons.spec.ts]: Remove

File: src/clr-icons/shapes/essential.umd.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  ClrShapeShrinkSVG,
+  ClrShapeTimesCircleSVG,
+  ClrShapeUsersSVG,
+  EssentialShapes,
+} from './essential';
+import { descriptorConfig } from '../utils/descriptor-config';
+
+export * from './essential';
+
+Object.defineProperty(EssentialShapes, 'edit', descriptorConfig(ClrShapePencilSVG));
```
Comment (gnomeontherun) [inline src/clr-icons/shapes/essential.umd.ts]: I believe we can remove all of these aliases from the umd files right?

File: src/dev/src/main.ts [hunk 9]
```diff
@@ -3,8 +3,8 @@
  * This software is released under MIT license.
  * The full license information can be found in LICENSE in the root directory of this project.
  */
-import '@clr/icons';
-import '@clr/icons/shapes/essential-shapes';
+import { ClarityIcons } from '@clr/icons';
+import { AllShapes } from '@clr/icons';
```
Comment (gnomeontherun) [inline src/dev/src/main.ts]: Can combine into one import.

File: src/ks-app/src/app/containers/iconography/icon-selection.component.ts [hunk 10]
```diff
@@ -6,15 +6,15 @@
...
-import { ChartShapes } from '@clr/icons/shapes/chart-shapes';
-import { CommerceShapes } from '@clr/icons/shapes/commerce-shapes';
-import { CoreShapes } from '@clr/icons/shapes/core-shapes';
-import { EssentialShapes } from '@clr/icons/shapes/essential-shapes';
-import { MediaShapes } from '@clr/icons/shapes/media-shapes';
-import { SocialShapes } from '@clr/icons/shapes/social-shapes';
-import { TechnologyShapes } from '@clr/icons/shapes/technology-shapes';
-import { TextEditShapes } from '@clr/icons/shapes/text-edit-shapes';
-import { TravelShapes } from '@clr/icons/shapes/travel-shapes';
+import { ChartShapes } from '@clr/icons/shapes/chart';
```
Comment (gnomeontherun) [inline src/ks-app/src/app/containers/iconography/icon-selection.component.ts]: Can we just move these imports up to `main.ts` like in the dev app and do `AllShapes` instead?

File: webpack.icons.config.js [hunk 11]
```diff
@@ -10,33 +10,31 @@ const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
...
-    'shapes/social-shapes.min': './src/clr-icons/shapes/social-shapes.ts',
-    'shapes/travel-shapes': './src/clr-icons/shapes/travel-shapes.ts',
-    'shapes/travel-shapes.min': './src/clr-icons/shapes/travel-shapes.ts',
-    'shapes/technology-shapes': './src/clr-icons/shapes/technology-shapes.ts',
-    'shapes/technology-shapes.min': './src/clr-icons/shapes/technology-shapes.ts',
-    'shapes/chart-shapes': './src/clr-icons/shapes/chart-shapes.ts',
-    'shapes/chart-shapes.min': './src/clr-icons/shapes/chart-shapes.ts',
-    'shapes/text-edit-shapes': './src/clr-icons/shapes/text-edit-shapes.ts',
-    'shapes/text-edit-shapes.min': './src/clr-icons/shapes/text-edit-shapes.ts',
+    index: './src/clr-icons/api.ts',
```
Comment (gnomeontherun) [inline webpack.icons.config.js]: Remove this line.

File: webpack.icons.config.js [hunk 12]
```diff
@@ -10,33 +10,31 @@ const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
...
+    'clr-icons-api': './src/clr-icons/api.ts',
```
Comment (gnomeontherun) [inline webpack.icons.config.js]: We don't want this line anymore.

File: webpack.icons.config.js [hunk 13]
```diff
@@ -10,33 +10,31 @@ const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
...
+    'clr-icons-base': './src/clr-icons/api-autoload.ts',
+    'clr-icons': './src/clr-icons/autoload.ts',
+    'clr-icons.min': './src/clr-icons/autoload.ts',
```
Comment (gnomeontherun) [inline webpack.icons.config.js]: Lines 15-17 lines should be replaced with

```
'clr-icons-all': './src/clr-icons/autoload-all.umd.ts',
'clr-icons-all.min': './src/clr-icons/autoload-all.umd.ts',
'clr-icons': './src/clr-icons/autoload.umd.ts',
'clr-icons.min': './src/clr-icons/autoload.umd.ts',

File: scripts/write-svg-icons.js (hunk 4 above)
Comment (jeeyun) [reply scripts/write-svg-icons.js]: We do.. because `text-edit` :)

File: src/clr-icons/element.ts [hunk 14]
```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { ClarityIconsApi } from './api';
+import { ShapeTemplateObserver } from './utils/shape-template-observer';
+
+export namespace ClarityIcon {
```
Comment (gnomeontherun) [inline src/clr-icons/element.ts]: I'm not happy with the namespace names I gave. Let's change this to `export namespace ElementModule`.

File: src/clr-icons/utils/shape-template-observer.ts [hunk 15]
```diff
@@ -5,51 +5,43 @@
  */
 import { ShapeTemplateObservables } from '../interfaces/icon-interfaces';
 
-export const changeHandlerCallbacks: ShapeTemplateObservables = {};
-
-export class ShapeTemplateObserver {
-  private static singleInstance: ShapeTemplateObserver;
-
-  private callbacks: ShapeTemplateObservables = changeHandlerCallbacks;
+export namespace ShapeTemplateObserver {
```
Comment (gnomeontherun) [inline src/clr-icons/utils/shape-template-observer.ts]: I'm not happy with the namespace names I gave. Let's change this to `export namespace ShapeTemplateObserverModule`

File: src/clr-icons/utils/shape-template-observer.ts [hunk 16]
```diff
@@ -5,51 +5,43 @@
...
+  export const changeHandlerCallbacks: ShapeTemplateObservables = {};
+
+  class ShapeTemplateObserverImplementation {
```
Comment (gnomeontherun) [inline src/clr-icons/utils/shape-template-observer.ts]: Let's rename this back to `ShapeTemplateObserver`.

File: src/clr-icons/api.ts [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { IconAlias, IconShapeSources } from './interfaces/icon-interfaces';
+import { ShapeTemplateObserver } from './utils/shape-template-observer';
+import { ClarityIcon } from './element';
+
+export namespace ClarityIconsApi {
+  export class ClarityIconsApiImplementation {
```
Comment (gnomeontherun) [inline src/clr-icons/api.ts]: Let's change this to
```
export namespace ApiModule {
  export class ClarityIconsApi {
```

File: src/clr-angular/data/datagrid/datagrid-page-size.spec.ts [hunk 18]
```diff
@@ -37,7 +37,8 @@ export default function(): void {
         it('displays a select with a default pageSize if no input is given', function() {
           const select = context.clarityElement.querySelector('select');
           expect(select).not.toBeNull();
-          expect(select.innerText).toMatch('0');
+          expect(select.options.length).toBe(1);
```
Comment (Shijir) [inline src/clr-angular/data/datagrid/datagrid-page-size.spec.ts]: Are these changes related to this PR?

File: src/clr-icons/shapes/all.umd.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { AllShapes } from './all';
+
+export * from './all';
```
Comment (Shijir) [inline src/clr-icons/shapes/all.umd.ts]: Do we need this export?

File: src/clr-angular/data/datagrid/datagrid-page-size.spec.ts (hunk 18 above)
Comment (jeeyun) [reply src/clr-angular/data/datagrid/datagrid-page-size.spec.ts]: No, but the build doesn't pass without it. It's already a part of master branch. I just added those fix for now, but when I rebased it should already be incorporated into it.

File: src/clr-icons/shapes/essential.umd.ts (hunk 8 above)
Comment (Shijir) [reply src/clr-icons/shapes/essential.umd.ts]: Yeah, right. It looks like we already created the aliases and assigned them to `EssentialShapes ` in the `essential`.ts file.

File: src/clr-icons/shapes/core.ts [hunk 20]
```diff
@@ -288,72 +288,108 @@ export const ClrShapeEvent = clrIconSVG(
...
             <path d="M34.29,17.53c-3.37-6.23-9.28-10-15.82-10a16.82,16.82,0,0,0-5.24.85L14.84,10a14.78,14.78,0,0,1,3.63-.47c5.63,0,10.75,3.14,13.8,8.43a17.75,17.75,0,0,1-4.37,5.1l1.42,1.42a19.93,19.93,0,0,0,5-6l.26-.48Z" class="clr-i-solid clr-i-solid-path-2"/>
             <path d="M4.87,5.78l4.46,4.46a19.52,19.52,0,0,0-6.69,7.29L2.38,18l.26.48c3.37,6.23,9.28,10,15.82,10a16.93,16.93,0,0,0,7.37-1.69l5,5,1.75-1.5-26-26Zm8.3,8.3a6.85,6.85,0,0,0,9.55,9.55l1.6,1.6a14.91,14.91,0,0,1-5.86,1.2c-5.63,0-10.75-3.14-13.8-8.43a17.29,17.29,0,0,1,6.12-6.3Z" class="clr-i-solid clr-i-solid-path-3"/>`
 );
-export const ClrShapeStepForward2 = clrIconSVG(
+export const ClrShapeStepForward2SVG = clrIconSVG(
   `<path d="M7.08,6.52a1.68,1.68,0,0,0,0,2.4L16.51,18,7.12,27.08a1.7,1.7,0,0,0,2.36,2.44h0L21.4,18,9.48,6.47A1.69,1.69,0,0,0,7.08,6.52Z" class="clr-i-outline clr-i-outline-path-1" /><path d="M26.49,5a1.7,1.7,0,0,0-1.7,1.7V29.3a1.7,1.7,0,0,0,3.4,0V6.7A1.7,1.7,0,0,0,26.49,5Z" class="clr-i-outline clr-i-outline-path-2" />`
 );
 
+export const ClrShapeUnknownStatus = { 'unknown-status': ClrShapeUnknownStatusSVG };
+export const ClrShapeHome = { home: ClrShapeHomeSVG };
```
Comment (Shijir) [inline src/clr-icons/shapes/core.ts]: I think the aliases should be assigned to their corresponding icons instead of the sets.

For example, currently we have this:
```
Object.defineProperty(CoreShapes, 'house', descriptorConfig(ClrShapeHomeSVG));
```
We users import individual shapes, their aliases would be missing. To have the default aliases, they have to import the whole sets.

So I think it should be like this instead: 
```
Object.defineProperty(ClrShapeHome, 'house', descriptorConfig(ClrShapeHomeSVG));
```
I think this would make sense to users. What do you think?

CC: @gnomeontherun

File: src/clr-icons/shapes/all.umd.ts (hunk 19 above)
Comment (gnomeontherun) [reply src/clr-icons/shapes/all.umd.ts]: Technically no, except to make the svg builder script work.

File: src/clr-icons/shapes/core.ts (hunk 20 above)
Comment (gnomeontherun) [reply src/clr-icons/shapes/core.ts]: This is a fair question, I even talked with @jeeyun about why we even bother having them. 

To me, it does makes sense that the alias would be tied to the shape rather than the set. I'd support that change.

File: src/clr-icons/shapes/all.umd.ts (hunk 19 above)
Comment (jeeyun) [reply src/clr-icons/shapes/all.umd.ts]: Are you asking for `all.umd.ts` specifically or `<setName>umd.ts` files in general? For each shape set, the export was necessary for our script that writes out the individual svg files.

File: src/clr-icons/shapes/essential.umd.ts (hunk 8 above)
Comment (jeeyun) [reply src/clr-icons/shapes/essential.umd.ts]: Yes. Just confirmed and will remove.

File: src/clr-icons/shapes/all.umd.ts (hunk 19 above)
Comment (Shijir) [reply src/clr-icons/shapes/all.umd.ts]: I see. That makes sense.

File: src/clr-icons/shapes/core.ts (hunk 20 above)
Comment (jeeyun) [reply src/clr-icons/shapes/core.ts]: It's an interesting idea. Depends on if we expect the aliases to be available when importing individual shape vs importing and adding a shape set. If we want both to work, we could do it by defining it twice.. so in your example above, once for `CoreShapes` and once for `ClrShapeHome`.

Comment (Shijir) [reply src/clr-icons/shapes/core.ts]: I don't think we need to define the aliases twice. Assigning the aliases to its corresponding icon object once could work. As we use the spread operator to dump individual shapes in the set object, the aliases would be dropped in just as well.

File: tslint.json [hunk 21]
```diff
@@ -38,6 +37,6 @@
     "semicolon": [true, "always", "ignore-bound-class-methods"],
     "switch-default": true,
     "triple-equals": [true, "allow-null-check"],
-    "variable-name": [true, "check-format", "allow-leading-underscore", "ban-keywords"]
+    "variable-name": [true, "check-format", "allow-leading-underscore", "allow-pascal-case", "ban-keywords"]
```
Comment (youdz) [inline tslint.json]: Any chance we could allow this only for icons? It saved me a couple times on our Angular components when declaring constants for providers.

File: src/clr-icons/shapes/core.ts [hunk 22]
```diff
@@ -342,6 +342,20 @@ export const ClrShapeEye = { eye: ClrShapeEyeSVG };
 export const ClrShapeEyeHide = { 'eye-hide': ClrShapeEyeHideSVG };
 export const ClrShapeStepForward2 = { 'step-forward-2': ClrShapeStepForward2SVG };
 
+Object.defineProperty(ClrShapeHome, 'house', descriptorConfig(ClrShapeHomeSVG));
```
Comment (Shijir) [inline src/clr-icons/shapes/core.ts]: Thank you, Jeeyun! :D Have we tested it by importing an individual shape? For example, if we import an individual shape, we expect both its icon name and aliases are added? 

I am a bit concerned that had the object been already exported, the alias would be actually missing because we are assigning the alias after the export.
```
export const ClrShapeHome = { home: ClrShapeHomeSVG };
Object.defineProperty(ClrShapeHome, 'house', descriptorConfig(ClrShapeHomeSVG));
```

File: tslint.json [hunk 23]
```diff
@@ -28,7 +28,6 @@
     "no-switch-case-fall-through": true,
     "no-this-assignment": [true, { "allow-destructuring": true }],
     "no-unused-expression": true,
-    "no-unused-variable": true,
```
Comment (youdz) [inline tslint.json]: Ok, this is a real problem for the Angular code. Without this, `tslint:fix` doesn't clean up unused imports anymore. 

I don't want to put 1.0 RC on hold for this, but I really don't agree with relaxing our TSLint rules for the `@clr/angular` just because they don't work for our icons. I believe having an extra tslint config in the `icons/` folder overrides the main config.

File: src/clr-icons/shapes/core.ts (hunk 22 above)
Comment (gnomeontherun) [reply src/clr-icons/shapes/core.ts]: The old way worked too where aliases were added after the setShapes was exported. This is because the Object.defineProperty method mutates via the reference so it trickles to the exported value as well.

File: tslint.json (hunk 23 above)
Comment (gnomeontherun) [reply tslint.json]: This is actually deprecated by tslint as a rule in general, it wants you to use the tsconfig flags for this instead.

File: tslint.json (hunk 21 above)
Comment (jeeyun) [reply tslint.json]: Yup makes sense and I was prob being too lazy. I'll change to allow only for icons.

File: src/clr-icons/shapes/core.ts (hunk 22 above)
Comment (Shijir) [reply src/clr-icons/shapes/core.ts]: You might be correct there. But also I think that could have been case due to this block of code. We had this in every set of icons file. We did all our alias assignments before this:
```ts
if (typeof window !== 'undefined' && window.hasOwnProperty('ClarityIcons')) {
    window.ClarityIcons.add(ChartShapes);
}
```
Now icons are separate and individually inserted so I think it's worth testing icon and its alias by importing by their individual host object.

Comment (Shijir) [reply src/clr-icons/shapes/core.ts]: Yesss! Awesome 👍
